File: src/main/scala/problems/simple/WordMachine.scala
package problems.simple

import cats.data.EitherT
import cats.implicits._
import cats.effect.IO
import scala.util.matching.Regex

object WordMachine {

  private val MIN = 0
  private val MAX = 0x7FFF_FFFF
  private val NumericPattern: Regex = "\\d+".r

  type Stack = List[Int]
  type Result[T] = Either[String, T]

  def main(args: Array[String]): Unit = {
    val wm = new WordMachine
    val result = wm.process("1 1 + 2 + DUP +")
    println(result)
  }

  class WordMachine {

    def process(input: String): Result[Int] =
      input
        .split(" ")
        .toList
        .foldM(
          List.empty[Int]
        )(
          (stack, token) => apply(stack, token)
        ).flatMap(pop).map(_._1)

    private def apply(stack: Stack, token: String): Result[Stack] = token match {
      case NumericPattern() => push(stack, token.toInt)
      case "POP" => pop(stack).map(_._2)
      case "DUP" => dup(stack)
      case "+" => add(stack)
      case "-" => sub(stack)
      case _ => Left(s"Unknown operation: $token")
    }

    private def push(stack: Stack, value: Int): Result[Stack] = {
      withinRange(value).map(validValue => validValue :: stack)
    }

    private def pop(stack: Stack): Result[(Int, Stack)] = {
      stack match {
        case head :: tail => Right((head, tail))
        case Nil => Left("Too few elements are available")
      }
    }

    private def dup(stack: Stack): Result[Stack] = {
      stack match {
        case head :: _ => Right(head :: stack)
        case Nil => Left("Too few elements are available")
      }
    }

    private def add(stack: Stack): Result[Stack] = {
      stack match {
        case a :: b :: tail => push(tail, a + b)
        case _ => Left("Too few elements are available")
      }
    }

    private def sub(stack: Stack): Result[Stack] = {
      stack match {
        case a :: b :: tail => push(tail, b - a)
        case _ => Left("Too few elements are available")
      }
    }

    private def withinRange(value: Int): Result[Int] = {
      if (value < MIN || value > MAX) Left("Number is out of integer range")
      else Right(value)
    }
  }
}

File: src/main/java/problemslab/WordMachine.java
package problemslab;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;
import java.util.function.Supplier;
import java.util.regex.Pattern;

public class WordMachine {
    private static final int MIN = 0;
    private static final int MAX = 0x7FFF_FFFF;
    private static final Pattern NUMERIC_PATTERN = Pattern.compile("\\d+");
    private final Stack<Integer> stack = new Stack<>();
    private static final Map<String, Supplier<Integer>> FUNCTIONS = new HashMap<>();
    {
        FUNCTIONS.put("POP", this::pop);
        FUNCTIONS.put("DUP", this::dup);
        FUNCTIONS.put("+", this::add);
        FUNCTIONS.put("-", this::sub);
    }

    public static void main(String[] args) {
        WordMachine wm = new WordMachine();
        System.out.println(wm.process("1 1 + 2 + DUP +"));
    }

    private int process(String s) {
        try {
            Arrays.stream(s.split(" ")).forEach(this::apply);
            return pop();
        } catch (IllegalArgumentException e) {
            return -1;
        }
    }

    private void apply(String token) {
        if (NUMERIC_PATTERN.matcher(token).matches()) {
            push(Integer.valueOf(token));
        } else {
            Supplier<Integer> function = FUNCTIONS.get(token);
            if (function != null) {
                function.get();
            } else {
                throw new IllegalArgumentException("Unknown operation: " + token);
            }
        }
    }

    private int push(int i) {
        withinRange(i);
        return stack.push(i);
    }

    private int pop() {
        return stack.pop();
    }

    private int dup() {
        hasElements(1);
        return push(stack.peek());
    }

    private int add() {
        hasElements(2);
        return push(stack.pop() + stack.pop());
    }

    private int sub() {
        hasElements(2);
        return push(stack.pop() - stack.pop());
    }

    private int hasElements(int i) {
        if(stack.size() < i){
            throw new IllegalArgumentException("Too few elements are available");
        }
        return i;
    }

    private int withinRange(int i){
        if(i < MIN || i > MAX){
            throw new IllegalArgumentException("Number is out of integer range");
        }
        return i;
    }
}
